aspnetcore/mvc/models/model-binding.md

```diff
@@ -1,146 +1,434 @@
+  ```
+
+* Form or query string data can be in one of the following formats:
+   
+  ```
+  selectedCourses=1050&selectedCourses=2000 
+  ```
+
+  ```
+  selectedCourses[0]=1050&selectedCourses[1]=2000
+  ```
+
+  ```
+  [0]=1050&[1]=2000
+  ```
+
+  ```
+  selectedCourses[a]=1050&selectedCourses[b]=2000&selectedCourses.index=a&selectedCourses.index=b
+  ```
+
+  ```
+  [a]=1050&[b]=2000&index=a&index=b
+  ```
+
+* The following format is supported only in form data:
+
+  ```
+  selectedCourses[]=1050&selectedCourses[]=2000
+  ```
+
+* For all of the preceding example formats, model binding passes an array of two items to the `selectedCourses` parameter:
+
+  * selectedCourses[0]=1050
+  * selectedCourses[1]=2000
+
+  Data formats that use subscript numbers (... [0] ... [1] ...) must ensure that they are numbered sequentially starting at zero. If there are any gaps in subscript numbering, all items after the gap are ignored. For example, if the subscripts are 0 and 2 instead of 0 and 1, the second item is ignored.
+
+## Dictionaries
+
+For `Dictionary` targets, model binding looks for matches to *parameter_name* or *property_name*. If no match is found, it looks for one of the supported formats without the prefix. For example:
+
+* Suppose the target parameter is a `Dictionary<string, string>` named `selectedCourses`:
+
+  ```csharp
+  public IActionResult OnPost(int? id, Dictionary<int, string> selectedCourses)
+  ```
+
+* The posted form or query string data can look like one of the following examples:
+
+  ```
+  selectedCourses[1050]=Chemistry&selectedCourses[2000]=Economics
+  ```
+
+  ```
+  [1050]=Chemistry&selectedCourses[2000]=Economics
+  ```
+
+  ```
+  selectedCourses[0].Key=1050&selectedCourses[0].Value=Chemistry&
+  selectedCourses[1].Key=2000&selectedCourses[1].Value=Economics
+  ```
+
+  ```
+  [0].Key=1050&[0].Value=Chemistry&[1].Key=2000&[1].Value=Economics
+  ```
+
+* For all of the preceding example formats, model binding passes a dictionary of two items to the `selectedCourses` parameter:
+
+  * selectedCourses["1050"]="Chemistry"
+  * selectedCourses["2000"]="Economics"
+
+## Special data types
+
+There are some special data types that model binding can handle.
+
+### IFormFile
+
+An uploaded file included in the HTTP request.  Also supported is `IEnumerable<IFormFile>` for multiple files.
+
+### CancellationToken
+
+Used to cancel activity in asynchronous controllers.
+
+### FormCollection
+
+Used to retrieve all the values from posted form data.
+
+## Input formatters
+
+Data in the request body can be in JSON, XML, or some other format. To parse this data, model binding uses an *input formatter* that is configured to handle a particular content type. By default, ASP.NET Core includes <xref:Microsoft.AspNetCore.Mvc.Formatters.JsonInputFormatter> and <xref:Microsoft.AspNetCore.Mvc.Formatters.JsonPatchInputFormatter> classes for handling JSON data. You can add other formatters for other content types.
```

Comment:
> By default, ASP.NET Core includes ...

This is not true in 3.0. We have a different set of formatter types. Could we instead say 

> By default, ASP.NET Core includes JSON based input formatters for handling JSON data.